### PR TITLE
Add widget themes

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | -

### CODE REVIEW

File: docs/guide/en/themes.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Themes
+
+Theme is named extra configuration of widget that merged with default configuration. Themes defined in
```
Comment (samdark) [inline docs/guide/en/themes.md:3]:
```suggestion
Theme is a named extra configuration of a widget that is merged with default configuration. Themes are defined in
```

File: docs/guide/en/themes.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+`WidgetFactory::initialize()`:
+
+```php
+\Yiisoft\Widget\WidgetFactory::initialize(
+    $container,
+     definitions: [
```
Comment (samdark) [inline docs/guide/en/themes.md:9]:
```suggestion
    $container,
    definitions: [
```

File: docs/guide/en/themes.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        MyWidget::class => [
+            '__construct()' => [
+                'name' => 'Base',
+            ],
+        ],
+    ],
+    themes: [
+        'colorize' => [
```
Comment (samdark) [inline docs/guide/en/themes.md:17]:
```suggestion
        'red-alert' => [
```

;)

File: docs/guide/en/themes.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            ],
+        ],
+    ],
+);
+```
+
+For a description of the configuration syntax, see the
+[Yii Definitions](https://github.com/yiisoft/definitions#arraydefinition) package documentation.
+
+To apply theme to widget call the `widget()` method with `theme` argument:
```
Comment (samdark) [inline docs/guide/en/themes.md:38]:
```suggestion
To apply a theme to the widget call the `widget()` method with `theme` argument:
```

File: docs/guide/en/themes.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+```php
+MyWidget::widget(theme: 'colorize');
```
Comment (samdark) [inline docs/guide/en/themes.md:41]:
```suggestion
MyWidget::widget(theme: 'red-alert');
```

File: docs/guide/en/widget-configuring.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Configuring the widget
+
+You can configure the widget when creating its instance, for example, the widget class must accept some ID when
+initializing the object.
```
Comment (samdark) [inline docs/guide/en/widget-configuring.md:4]:
```suggestion
You can configure the widget when creating its instance. For example, the widget class must accept some ID when
initializing the object.
```

File: docs/guide/en/widget-configuring.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+To set a value for the ID, you can pass it to the `widget()` method:
+
+```php
+<?= MyWidget::widget([
+    'id' => 'value',
+]) ?>
+```
+
+When you need extended configuration of widget (set properties or call methods) pass array definition via `config`
+parameter:
```
Comment (samdark) [inline docs/guide/en/widget-configuring.md:32]:
```suggestion
When you need extended configuration of a widget (to set properties or call methods) pass array definition via `config`
parameter:
```

File: docs/guide/en/widget-for-capturing-content.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+methods. These are wrapping widgets that mimic opening and closing HTML tags that wrap some content.
+They are used a bit differently:
+
+```php
+<?= MyWidget::widget()->begin() ?>
+    Content
+<?= MyWidget::end() ?>
+```
+
+For your widget to do this, you need override the parent `begin()` method and don't forget to call `parent::begin()`:
```
Comment (samdark) [inline docs/guide/en/widget-for-capturing-content.md:13]:
```suggestion
For your widget to do this, you need to override the parent `begin()` method. Don't forget to call `parent::begin()`:
```

File: src/WidgetFactory.php [hunk 9]
```diff
@@ -55,14 +80,74 @@ public static function initialize(
...
     {
         if (self::$factory === null) {
             throw new WidgetFactoryInitializationException(
                 'Widget factory should be initialized with WidgetFactory::initialize() call.',
             );
         }
 
+        $theme ??= self::$defaultTheme;
+
+        if ($theme !== null && isset(self::$themes[$theme])) {
```
Comment (samdark) [inline src/WidgetFactory.php:93]: array_key_exists?

File: src/WidgetFactory.php [hunk 10]
```diff
@@ -55,14 +80,74 @@ public static function initialize(
...
+                && isset(
+                    $config[ArrayDefinition::CLASS_NAME],
+                    self::$themes[$theme][$config[ArrayDefinition::CLASS_NAME]]
+                )
+            ) {
+                $config = ArrayDefinitionHelper::merge(
+                    self::$themes[$theme][$config[ArrayDefinition::CLASS_NAME]],
+                    $config
+                );
+            }
```
Comment (samdark) [inline src/WidgetFactory.php:105]:
```suggestion
            $className = $config[ArrayDefinition::CLASS_NAME] ?? null;
            if ($className !== null && isset(self::$themes[$theme][$className])) {
                $config = ArrayDefinitionHelper::merge(
                    self::$themes[$theme][$className],
                    $config
                );
            }
```

File: src/WidgetFactory.php [hunk 11]
```diff
@@ -55,14 +80,74 @@ public static function initialize(
...
+        }
+
         return self::$factory->create($config);
     }
+
+    /**
+     * @throws InvalidConfigException
+     */
+    private static function validateThemes(array $themes): void
```
Comment (samdark) [inline src/WidgetFactory.php:114]:
```suggestion
    private static function assertThemesStructure(array $themes): void
```

**Review by samdark — APPROVED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -3,8 +3,9 @@
 ## 2.1.0 under development
 
 - Bug #83: Fix merge constructor arguments with array definition configuration into `Widget::widget()` (@vjik)
-- Enh #84: Add protected method `Widget::getDefaultConfig()` that allow to create logic with internal default
-  configuration of widget (@vjik) 
+- Enh #84, #87: Add protected method `Widget::getThemeConfig()` that allow to create logic with internal theme
```
[the remark this answers is not in the document]
Comment (vjik) [reply CHANGELOG.md:6]: May be "implement a logic of configuring" ?

File: src/WidgetFactory.php (hunk 9 above)
Comment (vjik) [reply src/WidgetFactory.php:93]: Here it's equal.

File: src/WidgetFactory.php (hunk 10 above)
Comment (vjik) [reply src/WidgetFactory.php]: Replace `$className !== null` to `is_string($className)` and apply it.

File: src/WidgetFactory.php (hunk 11 above)
Comment (vjik) [reply src/WidgetFactory.php]: done

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (samdark) [reply CHANGELOG.md:6]: Yes.

[the remark this answers is not in the document]
Comment (vjik) [reply CHANGELOG.md]: done

**Pull request merged**